py/modstruct.c

```diff
@@ -132,17 +132,34 @@ STATIC mp_obj_t struct_calcsize(mp_obj_t fmt_in) {
 }
 MP_DEFINE_CONST_FUN_OBJ_1(struct_calcsize_obj, struct_calcsize);
 
-STATIC mp_obj_t struct_unpack(mp_obj_t fmt_in, mp_obj_t data_in) {
-    // TODO: "The buffer must contain exactly the amount of data required by the format (len(bytes) must equal calcsize(fmt))."
-    const char *fmt = mp_obj_str_get_str(fmt_in);
+STATIC mp_obj_t struct_unpack_from(mp_uint_t n_args, const mp_obj_t *args) {
+    // unpack requires that the buffer be exactly the right size.
+    // unpack_from requires that the buffer be "big enough".
+    // Since we implement unpack and unpack_from using the same function
+    // we relax the "exact" requirement, and only implement "big enough".
+    uint size = MP_OBJ_SMALL_INT_VALUE(struct_calcsize(args[0]));
```

Comment:
I can't help but think that it's inefficient to call struct_calcsize, and then calcsize_items below, and then parse the fmt a third time... maybe the check for "buffer too small" can be done in the final loop that parses fmt?